### PR TITLE
feat: add Health Component

### PR DESCRIPTION
**Pull request by JoenvandeVorle**
## Description

Added the Health component to the player and AbstractEnemy.
Also moved the Enemy and Player folders inside an Entity folder together with the Health Component (sorry for the 31 files changed -_-).

## Testability

There is a HealthComponent Scene that has a Player and Enemy with the new Health Component.

## Checklist
- [x] Set the proper pull request name
- [x] Merged main into your branch
- [x] Added a scene to the game for your changes

### CODE REVIEW

**Review by QuakeEye — CHANGES_REQUESTED** [review]
image
Something seems to be wrong with the player components

File: aplib.net-demo/Assets/Scripts/Entities/Health.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
+using UnityEngine;
+
+/// <summary>
+/// Health component that can be added to any GameObject.
+/// Keeps track of the GameObject's health and whether it is dead.
+/// When it is dead, it should fire an event to notify other components.
+/// </summary>
+public class HealthComponent : MonoBehaviour
+{
+    [SerializeField] private int _maxHealth = 100;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/Health.cs:10]:
```suggestion
    [SerializeField]
    private int _maxHealth = 100;
```

File: aplib.net-demo/Assets/Scripts/Entities/Health.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    public int Health { get; private set; }
+    public bool IsDead => Health <= 0;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/Health.cs:12]: Comments are missing.

File: aplib.net-demo/Assets/Scripts/Entities/Health.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    /// <summary>
+    /// Reduces the GameObject's health by the specified amount.
+    /// If the health reaches zero, the GameObject dies.
+    /// Healing can be done by passing a negative value.
+    /// </summary>
+    /// <param name="amount">The amount </param>
+    public void ReduceHealth(int amount)
+    {
+        Health -= amount;
+        if (Health <= 0)
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/Health.cs:28]:
```suggestion
        if (IsDead)
```

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]
Great work! I left a few comments

**Comment by JoenvandeVorle** [comment]
> image Something seems to be wrong with the player components

I can not reproduce this behaviour anymore. Can you check again on your end?

**Review by QuakeEye — APPROVED** [review]

**Review by QuakeEye — APPROVED** [review]
image

File: aplib.net-demo/Assets/Scripts/Entities/HealthComponent.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        Health = _maxHealth;
+    }
+    
+    /// <summary>
+    /// Reduces the GameObject's health by the specified amount.
+    /// If the health reaches zero, the GameObject dies.
+    /// Healing can be done by passing a negative value.
+    /// </summary>
+    /// <param name="amount">The amount </param>
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/HealthComponent.cs:49]:
```suggestion
    /// <param name="amount">The amount of health to be subtracted.</param>
```

File: aplib.net-demo/Assets/Scripts/Entities/HealthComponent.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    public void ReduceHealth(int amount)
+    {
+        Health -= amount;
+        if (amount >= 0)
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/HealthComponent.cs:53]: Nit
```suggestion
        Health -= amount;
        
        if (amount >= 0)
```

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]
Looks great! One tiny problem with an XML summary.

**Review by QuakeEye — APPROVED** [review]

**Pull request merged**